### PR TITLE
 Add GetTool API and ListTools API

### PR DESCRIPTION
**Pull request by jackiehanyang**
### Description
This pr moves GetTool and ListTools API from feature_dev branch. All code should be identical except this pr removes the hardcoded internal tools part. Also, renaming all `externalTools` attribute as `toolMetadataList` to reduce confusion as we don't differentiate internal or external tools when building GetTool and ListTools result. Instead, we pass internal tools and external tools together as a whole tool factories to these two rest APIs.

Had to close the [original pr](https://github.com/opensearch-project/ml-commons/pull/1764) due to too much merge conflict, and wanted to make sure all code is from feature_dev branch and it don't get messed up during resolving merge conflicts. You can find references in the original pr conversations.

**Note**: Build failure doesn't seem to be related to this commit and has failed on previous commit with same error: https://github.com/opensearch-project/ml-commons/pull/1800. Root cause in https://github.com/opensearch-project/ml-commons/pull/1816/files#r1437937707

### Testing
```
GET http://localhost:9200/_plugins/_ml/tools/MLModelTool
{
	"name": "MLModelTool",
	"description": "Use this tool to run any model.",
	"type": "MLModelTool",
	"version": "undefined"
}
```
```
GET http://localhost:9200/_plugins/_ml/tools
{
	"name": "MLModelTool",
	"description": "Use this tool to run any model.",
	"type": "MLModelTool",
	"version": "undefined"
}{
	"name": "AgentTool",
	"description": "Use this tool to run any agent.",
	"type": "AgentTool",
	"version": "undefined"
}{
	"name": "CatIndexTool",
	"description": "Use this tool to get index information.",
	"type": "CatIndexTool",
	"version": "undefined"
}
```
 
### Issues Resolved
related issues - https://github.com/opensearch-project/ml-commons/issues/1802
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ylwu-amzn** [comment]
https://github.com/opensearch-project/ml-commons/actions/runs/7359436543/job/20034209306?pr=1818
```
* What went wrong:
Execution failed for task ':opensearch-ml-plugin:jacocoTestCoverageVerification'.
> Rule violated for class org.opensearch.ml.action.tools.ListToolsTransportAction: lines covered ratio is 0.2, but expected minimum is 0.8
  Rule violated for class org.opensearch.ml.action.tools.GetToolTransportAction: lines covered ratio is 0.1, but expected minimum is 0.8
  Rule violated for class org.opensearch.ml.rest.RestMLGetToolAction: lines covered ratio is 0.5, but expected minimum is 0.8
  Rule violated for class org.opensearch.ml.rest.RestMLListToolsAction: lines covered ratio is 0.5, but expected minimum is 0.8
```

File: common/src/main/java/org/opensearch/ml/common/ToolMetadata.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import static org.opensearch.core.xcontent.XContentParserUtils.ensureExpectedToken;
+
+
+public class ToolMetadata implements ToXContentObject, Writeable {
+
+    public static final String TOOL_NAME_FIELD = "name";
+    public static final String TOOL_DESCRIPTION_FIELD = "description";
+
+    @Getter
+    private String name;
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/ToolMetadata.java:28]: Jing did a refactor and changed name to type([code link](https://github.com/opensearch-project/ml-commons/commit/90d9b310811c682f55ee5415d53ad0be8f671156#diff-f3f27d4e953486c2798c6edef8641779b3211dabbe5374c5bfe1737b3aab2f25)). Change to type?

Comment (jackiehanyang) [reply common/src/main/java/org/opensearch/ml/common/ToolMetadata.java:28]: Jing's change still kept `name`, just added a new field `type`. Do we want to include `type` in ToolMetadata and show it in GetTool and ListTools response?

Comment (dhrubo-os) [reply common/src/main/java/org/opensearch/ml/common/ToolMetadata.java:28]: I think we should show `type` & `version` both.

Comment (jngz-es) [reply common/src/main/java/org/opensearch/ml/common/ToolMetadata.java:28]: Yeah, we need both type and name.

Comment (jackiehanyang) [reply common/src/main/java/org/opensearch/ml/common/ToolMetadata.java:28]: Should we add a default version each tool? Currently it's set to null - https://github.com/opensearch-project/ml-commons/blob/8352e69ae4edd2b615933ff2947d61d1d9546740/ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/AgentTool.java#L78

Comment (jackiehanyang) [reply common/src/main/java/org/opensearch/ml/common/ToolMetadata.java:31]: Added type and version information in the latest revision

File: plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+        List<ToolMetadata> toolsList = mlToolsGetRequest.getToolMetadataList();
+
+        try {
+            listener.onResponse(MLToolsListResponse.builder().toolMetadata(toolsList).build());
+        } catch (Exception e) {
+            log.error("Failed to get tools list", e);
+            listener.onFailure(e);
+        }
+    }
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java:46]: Do we really need this transport layer action if it's only returning the toolsList from the ListToolsRequest? It seems like the response is already included in the Request, so this response can be returned right away when the Request is constructed?

File: plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            ToolMetadata theTool = toolsList
+                .stream()
+                .filter(tool -> tool.getName().equals(toolName))
+                .findFirst()
+                .orElseThrow(NoSuchElementException::new);
+            listener.onResponse(MLToolGetResponse.builder().toolMetadata(theTool).build());
+        } catch (Exception e) {
+            log.error("Failed to get tool", e);
+            listener.onFailure(e);
+        }
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java:53]: Why not just doing this in the rest layer and return the response right away? Or whoever calls this action can do it directly from the client side?

File: plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java (hunk 2 above)
Comment (jackiehanyang) [reply plugin/src/main/java/org/opensearch/ml/tools/ListToolsTransportAction.java:46]: Simply followed the current ml-commons API structure: from rest layer to transport layer. If we can guarantee that there won't be use case of calling this transport action directly in the future, then we can talk about a plan to only leave the rest layer action and remove this one

**Review by Zhangxunmt — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java (hunk 3 above)
Comment (jackiehanyang) [reply plugin/src/main/java/org/opensearch/ml/tools/GetToolTransportAction.java:53]: same as above comment, we can return the response right away from the client side. But will it be a short-term solution and make it hard for future iterations?

File: common/src/test/java/org/opensearch/ml/common/transport/tools/MLToolGetRequestTests.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            }
+        };
+        MLToolGetRequest.fromActionRequest(actionRequest);
+    }
+
+    @Test
+    public void validate_Exception_NullToolName() {
+        MLToolGetRequest mlToolGetRequest = MLToolGetRequest.builder().build();
+        ActionRequestValidationException exception = mlToolGetRequest.validate();
+        assertEquals("Validation Failed: 1: Tool name can't be null;", exception.getMessage());
```
[the remark this answers is not in the document]
Comment (jackiehanyang) [reply common/src/test/java/org/opensearch/ml/common/transport/tools/MLToolGetRequestTests.java:95]: It's validating the tool name in the get request cannot be null, not the name of the actual tool cannot be null. So it's different than the code example you pasted and it doesn't apply in this case.

[the remark this answers is not in the document]
Comment (mingshl) [reply common/src/test/java/org/opensearch/ml/common/transport/tools/MLToolGetRequestTests.java:95]: Fair enough

**Pull request merged**